### PR TITLE
Add information on memory and network limits

### PR DESCRIPTION
**Pull request by iHiD**
We've discovered that there are important differences regarding network settings in different languages. This PR adds some docs around this, and also about memory usage.

The same applies to test runners and I'll copy this there too once it's refined/approved.

### CODE REVIEW

**Comment by NobbZ** [comment]
The `max_memory` configured, does that correspond to the limit the container has or is it also used by the scheduler to determine on what host there is still enough memory? Or only the latter?

In terms of k8s speak: Does it correspond to "request" or "limit" or both?

Having a "limit" is important for me, as then the VM will pick it up automatically and assume it as a system limit. If it is not such a limit the runtime will assume the physical memory of the host as upper limit.

If it is just a "request", then the Erlang runtime won't even know about it and use as much memory as it can before getting killed. In such a scenario I still could use (and would want to) command line magic to limit the runtimes memory.

**Comment by iHiD** [comment]
Thanks @NobbZ. So the simplest answer is that I don't know what is right/wrong and am really open to the knowledge of others. Currently I'm setting the `--memory` flag documented here: https://docs.docker.com/config/containers/resource_constraints/#limit-a-containers-access-to-memory

I would happily set all the flags if I understood more about what helps us here, and what sensible defaults would be.

What do you think?

(Also cc @bitfield as he literally wrote the book on this stuff)

**Comment by NobbZ** [comment]
Okay, that flag corresponds directly to the `limit` in k8s. As I have no clue how actual orchestration is done, I just assume that you will at least consider the memory limit to avoid overcommitment.

At least I know now, that from the Erlang side I do not need to do much, except for checking that I use the version that already speaks cgroups ;) I'm not sure if that feature was introduced with 22 or 23.

**Comment by iHiD** [comment]
> As I have no clue how actual orchestration is done

Docker is shelled out to here: https://github.com/exercism/tooling-invoker/blob/main/lib/tooling_invoker/exec_docker.rb#L154-L163

With the config read from a JSON file here: https://github.com/exercism/tooling-invoker/blob/main/tools.json

**Comment by verdammelt** [comment]
Just to clarify - the default, if we do not edit the tooling-websever tools.json file is: network=none and ~no limit to memory usage~ memory=3GB?

File: docs/docker.md [hunk 1]
```diff
@@ -18,6 +18,30 @@ Our analyzers and representers are deployed as docker images.
...
+
+### Network
+
+Tools run without access to the internet. There are two different configurations you can use:
+1. `none`. This disables the networking device inside the container.
+2. `internal`. This adds a networking device inside the container but the network has no access to anything external.
+
+Different languages perform better/worse with different configurations (e.g. Ruby is 2x faster with `none`. Elixir is `12x` faster with `internal`.
+
+You can experiement locally by using the `--network` flag when running your docker. `--network none` is supported by default.
```
Comment (ErikSchierboom) [inline docs/docker.md]:
```suggestion
You can experiment locally by using the `--network` flag when running your docker. `--network none` is supported by default. 
```

**Review by ErikSchierboom — APPROVED** [review]
Very nice!

**Comment by iHiD** [comment]
> Just to clarify - the default, if we do not edit the tooling-websever tools.json file is: network=none and no limit to memory usage memory=3GB?

This is *currently* correct, but these defaults may turn out to be very bad so are subject to change.

**Pull request merged**